### PR TITLE
Openlibrary tests

### PR DESCRIPTION
**Pull request by mnmallea**
## Summary
- Added tests for OpenlibraryService

## Trello card
[Link](https://trello.com/c/AQZOlu)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -25,6 +25,7 @@ gem 'redis', '~> 4.0'
 # gem 'rack-cors', '~>5.2.7'
 gem 'active_model_serializers', '~> 0.10.0'
 gem 'devise_token_auth'
+gem 'httparty'
```
Comment (lucasVoboril) [inline Gemfile]: This gem was installed in the previous PR. You should change your base branch :wink:

File: Gemfile [hunk 2]
```diff
@@ -56,6 +57,7 @@ group :test do
   gem 'rails-controller-testing'
   gem 'rspec-mocks'
   gem 'shoulda-matchers'
+  gem 'webmock'
```
Comment (lucasVoboril) [inline Gemfile]: Tag gem version
```suggestion
  gem 'webmock', '~> 3.5.1'
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -14,6 +14,11 @@ def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:sign_up, keys: %i[first_name last_name locale])
   end
 
+  def render_errors(message, status)
```
Comment (lucasVoboril) [inline app/controllers/application_controller.rb]: Also, this comes from the previous PR

File: spec/support/fixtures/openlibrary/response_success.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        "BQ9265.6 .T7313 1994"
+      ]
+    },
+    "publish_places": [
+      {
+        "name": "New York"
+      }
+    ]
+  }
+}
```
Comment (lucasVoboril) [inline spec/support/fixtures/openlibrary/response_success.json]: Add empty line at end of file. If you are using vscode you can add this settings 
`  "files.insertFinalNewline": true`

File: spec/support/mocks_and_stubbs/openlibrary_mock.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  def successfuly_book_get(isbn)
+    stub_request(:get, "#{BASE_URI}/books")
+      .with(
+        query: { bibkeys: "ISBN:#{isbn}", format: 'json', jscmd: 'data' }
+      )
+      .to_return(
+        status: 200,
+        headers: { 'Content-Type' => 'application/json' },
+        body: File.read('./spec/support/fixtures/openlibrary/response_success.json')
```
Comment (lucasVoboril) [inline spec/support/mocks_and_stubbs/openlibrary_mock.rb]:
```suggestion
        body: Rails.root.join('spec/support/fixtures/openlibrary/response_success.json')
```

File: spec/support/fixtures/openlibrary/response_success.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
+{
```
Comment (lucasVoboril) [inline spec/support/fixtures/openlibrary/response_success.json]: Move this file to 
> spec/support/fixtures/stub/openlibrary/responses/success.json

File: spec/support/mocks_and_stubbs/openlibrary_mock.rb (hunk 5 above)
Comment (lucasVoboril) [reply spec/support/mocks_and_stubbs/openlibrary_mock.rb]: > spec/support/fixtures/stub/openlibrary/responses/success.json

File: spec/support/mocks_and_stubbs/openlibrary_mock.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
+module OpenlibraryMock
+  BASE_URI = 'https://openlibrary.org/api'.freeze
```
Comment (lucasVoboril) [inline spec/support/mocks_and_stubbs/openlibrary_mock.rb]: This base_uri should be an env variable

File: spec/support/mocks_and_stubbs/openlibrary_mock.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        query: { bibkeys: "ISBN:#{isbn}", format: 'json', jscmd: 'data' }
+      )
+      .to_return(
+        status: 200,
+        headers: { 'Content-Type' => 'application/json' },
+        body: '{}'
+      )
+  end
+
+  def internal_server_error_book_get(isbn)
```
Comment (lucasVoboril) [inline spec/support/mocks_and_stubbs/openlibrary_mock.rb]: :warning: Warning: this not be an internal error. This is an openlibrary error
```suggestion
  def openlibrary_server_error_book_get(isbn)
```

File: spec/rails_helper.rb [hunk 9]
```diff
@@ -9,6 +9,7 @@
 Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
 require 'wor/paginate/rspec'
 require 'validate_url/rspec_matcher'
+require 'webmock/rspec'
```
Comment (lucasVoboril) [inline spec/rails_helper.rb:12]: You can add this below:
```rb
  config.include Devise::Test::ControllerHelpers, type: :controller
  config.include Response::JSONParser, type: :controller
  config.include OpenlibraryMock
```

File: spec/services/openlibrary_service_spec.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
+require 'rails_helper'
+
+describe OpenlibraryService do
+  include OpenlibraryMock
```
Comment (lucasVoboril) [inline spec/services/openlibrary_service_spec.rb]: We can omit to do this include :wink:

File: spec/services/openlibrary_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  let(:service) { OpenlibraryService.new }
+  let(:uri) { 'https://openlibrary.org/api/books' }
```
Comment (lucasVoboril) [inline spec/services/openlibrary_service_spec.rb]: :disappointed: env variable :pray:

File: spec/services/openlibrary_service_spec.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      end
+
+      it 'responses with a Hash' do
+        is_expected.to be_a Hash
+      end
+
+      it 'responses with a Hash with expected keys' do
+        expected_keys = %i[isbn title subtitle number_of_pages authors]
+        is_expected.to include(*expected_keys)
+      end
```
Comment (lucasVoboril) [inline spec/services/openlibrary_service_spec.rb:31]: We can test status code be :ok

**Review by lucasVoboril — CHANGES_REQUESTED** [review]
This PR contains previous PRs, you should change base branch :wink:

File: spec/services/openlibrary_service_spec.rb (hunk 12 above)
Comment (mnmallea) [reply spec/services/openlibrary_service_spec.rb:31]: I haven't got access to the http response outside the service object

File: app/services/openlibrary_service.rb [hunk 13]
```diff
@@ -13,7 +13,7 @@ def search_by_isbn(isbn)
 
     return parse_response_book(response[isbn_string], isbn) if response.success?
 
-    raise CustomErrors::BookNotFoundError, response.response
+    raise CustomErrors::BookNotFoundError, 'API request failed'
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: What API fails?
```suggestion
    raise CustomErrors::BookNotFoundError, 'Openlibrary API request failed'
```

File: spec/support/mocks_and_stubbs/openlibrary_mock.rb [hunk 14]
```diff
@@ -0,0 +1,39 @@
+module OpenlibraryMock
+  BASE_URI = ENV['OPEN_LIBRARY_URI']
```
Comment (lucasVoboril) [inline spec/support/mocks_and_stubbs/openlibrary_mock.rb]:
```suggestion
  BASE_URI = Rails.application.secrets.open_library_uri
```

File: spec/services/openlibrary_service_spec.rb [hunk 15]
```diff
@@ -0,0 +1,72 @@
+require 'rails_helper'
+
+describe OpenlibraryService do
+  let(:service) { OpenlibraryService.new }
+  let(:uri) { "#{ENV['OPEN_LIBRARY_URI']}/books" }
```
Comment (lucasVoboril) [inline spec/services/openlibrary_service_spec.rb]:
```suggestion
  let(:uri) { "#{Rails.application.secrets.open_library_uri}/books" }
```

File: spec/services/openlibrary_service_spec.rb [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+  describe '#search_by_isbn' do
+    subject { service.search_by_isbn(isbn) }
```
Comment (lucasVoboril) [inline spec/services/openlibrary_service_spec.rb]: Name subject
```suggestion
    subject(:service) { service.search_by_isbn(isbn) }
```

**Review by lucasVoboril — CHANGES_REQUESTED** [review]

File: spec/services/openlibrary_service_spec.rb (hunk 16 above)
Comment (mnmallea) [reply spec/services/openlibrary_service_spec.rb]: `search_by_isbn` returns a book, not a service :thinking:

**Review by lucasVoboril — APPROVED** [review]

**Pull request merged**